### ISSUE TITLE
Does not seems to export anything to exporter

### ISSUE THREAD

**Issue by RemiKalbe** [davidB/axum-tracing-opentelemetry#50]
Tldr:
Everything that happens before axum starts are sent to my exporter, everything after doesn't.

```rust
#[tokio::main]
async fn main() {
    init_tracer();

    let state = state::AppState::default(); // <--- This is sent to Grafana Tempo
    state.start_ws_observer_task(); // <--- This is sent to Grafana Tempo
    let config = state.config();  // <--- This is sent to Grafana Tempo

// After that, nothing is sent
    let addr = SocketAddr::from((config.env.listening_ip, config.env.listening_port));
    info!("Listening on: {}", addr);
    let app = Router::new()
        .nest("/api/v1", handlers::router())
        .with_state(state.clone())
        .layer(opentelemetry_tracing_layer());
    axum::Server::bind(&addr)
        .serve(app.into_make_service_with_connect_info::<SocketAddr>())
        .with_graceful_shutdown(shutdown_signal(state))
        .await
        .unwrap();
}
```

```rust
fn init_tracer() {
    let env = config::ENV::new();

    let env_filter = EnvFilter::try_from_default_env()
        .or_else(|_| EnvFilter::try_new("info,axum_tracing_opentelemetry=info"))
        .unwrap();

    let exporter = opentelemetry_otlp::new_exporter()
        .tonic()
        .with_endpoint(env.opentelemetry_endpoint.clone());

    let config =
        opentelemetry::sdk::trace::config().with_resource(opentelemetry::sdk::Resource::new(vec![
            opentelemetry::KeyValue::new("service.name", "gateway"),
            opentelemetry::KeyValue::new("service.namespace", "cloud-conference"),
        ]));

    let tracer = opentelemetry_otlp::new_pipeline()
        .tracing()
        .with_exporter(exporter)
        .with_trace_config(config)
        .install_batch(opentelemetry::runtime::Tokio)
        .expect("Failed to create opentelemtry pipeline");

    let telemetry = tracing_opentelemetry::layer().with_tracer(tracer.clone());

    let formatting_layer = BunyanFormattingLayer::new("gateway".to_string(), std::io::stdout);

    axum_tracing_opentelemetry::init_propagator().unwrap();

    let subscriber = Registry::default()
        .with(env_filter)
        .with(telemetry)
        .with(JsonStorageLayer)
        .with(formatting_layer);

    tracing::subscriber::set_global_default(subscriber)
        .expect("Failed to install `tracing` subscriber.");
}
```

Here is an exemple of a log, as you can see the `trace_id` exists, but the trace is not sent to Tempo for some reason, **it was working fine when I used actix web with the exact same opentelemetry/tracing config**
```rust
{"v":0,"name":"gateway","msg":"[HTTP REQUEST - START]","level":30,"hostname":"gateway-0","pid":1,"time":"2023-03-04T18:16:58.580886678Z","target":"axum_tracing_opentelemetry::middleware::trace_extractor","line":161,"file":"/usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/axum-tracing-opentelemetry-0.10.0/src/middleware/trace_extractor.rs","http.method":"GET","http.route":"","http.flavor":"1.1","http.target":"/api/v1/private/liveness","http.scheme":"HTTP","http.client_ip":"100.64.4.115:45702","http.user_agent":"kube-probe/1.24","otel.kind":"server","trace_id":"a7d4105269dd7e48a25d7849e9df333d","otel.name":"GET ","http.host":"100.64.4.172:8080"}
{"v":0,"name":"gateway","msg":"[HTTP REQUEST - END]","level":30,"hostname":"gateway-0","pid":1,"time":"2023-03-04T18:16:58.581061857Z","target":"axum_tracing_opentelemetry::middleware::trace_extractor","line":161,"file":"/usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/axum-tracing-opentelemetry-0.10.0/src/middleware/trace_extractor.rs","elapsed_milliseconds":0,"http.method":"GET","http.route":"","http.flavor":"1.1","http.target":"/api/v1/private/liveness","http.scheme":"HTTP","http.client_ip":"100.64.4.115:45702","http.user_agent":"kube-probe/1.24","otel.kind":"server","trace_id":"a7d4105269dd7e48a25d7849e9df333d","otel.name":"GET ","http.host":"100.64.4.172:8080","http.status_code":"204","otel.status_code":"OK"}
```
Here you can see, I do receive traces, but only about things that happens before axum starts, or about traces I create manually in my websocket task.
<img width="958" alt="CleanShot 2023-03-04 at 19 18 46@2x" src="https://user-images.githubusercontent.com/8604600/222922363-e84b8945-1389-4d2f-a49f-ae6afd9c2413.png">
If I search by trace_id, with the one that was in the log I provided, here is what happens:
<img width="953" alt="CleanShot 2023-03-04 at 19 19 18@2x" src="https://user-images.githubusercontent.com/8604600/222922396-a8e3f322-08df-4e8e-8a5c-239b2e62d5bc.png">

**Comment by davidB** [comment]
Into the route of axum, Did you add the layer ` .layer(opentelemetry_tracing_layer())` ? 

see README, the crate provide 2 axum layers:
-  1 to create trace (new or from remote) + span
- 1 (optional) to output trace info into the header of the response  (useful only for debug, dev)

**Comment by davidB** [comment]
From your `init_tracer` above, you didn't define explicitly the sampler. I don't know what is the expected behavior in this case. 

(eg in version before 0.10 the crate define the sampler to `Always` when you use the helper from the crate for the setup, for version 0.10, the value is based on the environment with default following Opentelementry spec)

The sampler define what is exported or not,
As your exporter seams to be correctly configured as you receive some trace, I only see the sampler as "blocker".

**Comment by davidB** [comment]
The default behavior is `parentbased(always_on)`  via opentelemetry crate (like in your case) or via helper of this crate.  (https://docs.rs/opentelemetry/0.18.0/opentelemetry/sdk/trace/enum.Sampler.html#variant.ParentBased). Trace created by this crate are flagged as remote (and not root iirc), so except if the caller create a trace to report, nothing will be exported (I guess)

You can try with a w3c tracecontext like `traceparent: 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01` (as header) to see if it is reported (the flag `01`  at end)

**Comment by RemiKalbe** [comment]
Indeed it works with
```rust
.with_sampler(opentelemetry::sdk::trace::Sampler::AlwaysOn);
```
Thank you for the detailed answer 😄.

I'm just a bit confused on why it was working fine when I used [tracing-actix-web](https://github.com/LukeMathWalker/tracing-actix-web) without setting a sampler.
